### PR TITLE
fix: close semantics

### PR DESCRIPTION
**Pull request by ronag**
https://github.com/mcollina/undici/issues/175

### CODE REVIEW

File: lib/client.js [hunk 1]
```diff
@@ -621,17 +644,31 @@ class Client extends EventEmitter {
...
+        callback(err, null)
+      }
+    }
+
     this[kClosed] = true
+    this[kOnDestroyed].push(onDestroyed)
+
+    for (const request of this[kQueue].slice(this[kInflight])) {
+      if (request.signal) {
+        request.signal.on('error', onDestroyed)
```
Comment (ronag) [inline lib/client.js:664]: should this still wait for the socket to be destroyed before invoking the callback?

**Comment by ronag** [comment]
I'm struggling with this. It's rather complicated.

I don't think the semantics of:

```js
const requests = makeLotsOfRequests()
await Promise.all(requests)
```

Is correct either since it doesn't actually wait for the responses.

We have to wait for the response (if exists) to complete as well.

**Comment by ronag** [comment]
@mcollina: Much of the difficult here lies with propagating the errors... would it be an option that `close(callback)` does not propagate errors? i.e. it just closes the client for further requests and waits for all requests to complete, whether they error or not.

Essentially it would be:

```js
Promise.allSettled(...)
  .then(() => callback(null))
  .catch((err) => callback(err))
```

**Comment by mcollina** [comment]
ok for me

**Comment by ronag** [comment]
this is ready for review

**Pull request merged**